adapters/ix/ix.go

```diff
@@ -1,267 +1,189 @@
+		request.Imp = imps[iImp : iImp+1]
+		if request.Site != nil {
+			setSitePublisherId(request, iImp)
 		}
-	}
 
-	ch := make(chan ixBidResult)
-	for _, request := range requests {
-		go func(bidder *pbs.PBSBidder, request callOneObject) {
-			result, err := a.callOne(ctx, request.requestJSON)
-			result.Request = &request
-			result.Error = err
-			if result.Bid != nil {
-				result.Bid.BidderCode = bidder.BidderCode
-				result.Bid.BidID = bidder.LookupBidID(result.Bid.AdUnitCode)
-				result.Bid.Width = request.width
-				result.Bid.Height = request.height
-
-				if result.Bid.BidID == "" {
-					result.Error = &errortypes.BadServerResponse{
-						Message: fmt.Sprintf("Unknown ad unit code '%s'", result.Bid.AdUnitCode),
+		if request.Imp[0].Banner != nil {
+			banner := *request.Imp[0].Banner
+			request.Imp[0].Banner = &banner
+			formats := getBannerFormats(&banner)
+			for iFmt := range formats {
+				banner.Format = formats[iFmt : iFmt+1]
+				banner.W = openrtb.Uint64Ptr(banner.Format[0].W)
+				banner.H = openrtb.Uint64Ptr(banner.Format[0].H)
+				if requestData, err := createRequestData(a, request, &headers); err == nil {
+					if iFmt == 0 {
+						requests = append(requests, requestData)
+					} else {
+						multiSizeRequests = append(multiSizeRequests, requestData)
 					}
-					result.Bid = nil
+				}
+				if len(multiSizeRequests) == cap(multiSizeRequests) {
+					break
 				}
 			}
-			ch <- result
-		}(bidder, request)
+		} else if requestData, err := createRequestData(a, request, &headers); err == nil {
+			requests = append(requests, requestData)
+		}
 	}
+	request.Imp = imps
 
-	bids := make(pbs.PBSBidSlice, 0)
-	for i := 0; i < len(requests); i++ {
-		result := <-ch
-		if result.Bid != nil && result.Bid.Price != 0 {
-			bids = append(bids, result.Bid)
-		}
+	return append(requests, multiSizeRequests...), errs
+}
 
-		if req.IsDebug {
-			debug := &pbs.BidderDebug{
-				RequestURI:   a.URI,
-				RequestBody:  result.Request.requestJSON.String(),
-				StatusCode:   result.StatusCode,
-				ResponseBody: result.ResponseBody,
-			}
-			bidder.Debug = append(bidder.Debug, debug)
-		}
-		if result.Error != nil {
-			err = result.Error
+func setSitePublisherId(request *openrtb.BidRequest, iImp int) {
+	if iImp == 0 {
+		// first impression - create a site and pub copy
+		site := *request.Site
+		if site.Publisher == nil {
+			site.Publisher = &openrtb.Publisher{}
+		} else {
+			publisher := *site.Publisher
+			site.Publisher = &publisher
 		}
+		request.Site = &site
 	}
 
-	if len(bids) == 0 {
-		return nil, err
-	}
-	return bids, nil
-}
-
-func (a *IxAdapter) callOne(ctx context.Context, reqJSON bytes.Buffer) (ixBidResult, error) {
-	var result ixBidResult
+	// 'adapters/bidder.go': Bidder implementations may safely assume that this
+	// JSON has been validated by their static/bidder-params/{bidder}.json file.
+	var bidderExt adapters.ExtImpBidder
+	var ixExt openrtb_ext.ExtImpIx
+	json.Unmarshal(request.Imp[0].Ext, &bidderExt)
```

Comment:
It is safe to assume the json is already validation, but you still need to bubble up the errors. There are still other situations where an error can occur.